### PR TITLE
Issue #ED-76 feat: Native sign-in in Sunbird mobile app

### PR DESCRIPTION
**Pull request by vpPavithra**

### CODE REVIEW

File: src/app/routes/keycloakSignInRoutes.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+      if (item['key'] == 'redirect_uri') {
+        redirect_uri = item['value'];
+      }
+      queryArr.push((Object.values(item)))
+    });
+    let reqOption = getRequestOptions(config.host+config.path, getHeaders(), qs.stringify(Object.fromEntries(queryArr)));
+    request(reqOption, ((err, response, data) => {
+      if (!err && response.statusCode == 200) {
+        logger.info({msg: 'keyclaok Signin route: initialized ' + reqOption.method + reqOption.url})
+        var htmlObject = parser.parseFromString(data, 'text/html');
```
Comment (rajeevsathish) [inline src/app/routes/keycloakSignInRoutes.js:28]: var cannot be use

**Review by rajeevsathish — CHANGES_REQUESTED** [review]

File: src/app/routes/keycloakSignInRoutes.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
+const bodyParser = require('body-parser');
+const envHelper = require('../helpers/environmentVariablesHelper.js');
+const request = require('request-promise');
+var qs = require('querystring');
```
Comment (rajeshkumaravel) [inline src/app/routes/keycloakSignInRoutes.js:4]: replace `var` with `const` or `let` accordingly

File: src/app/routes/keycloakSignInRoutes.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+var DOMParser = require('xmldom').DOMParser;
+const parser = new DOMParser();
+const { logger } = require('@project-sunbird/logger');
+
+module.exports = (app) => {
+  app.post('/keycloak/login', bodyParser.json(), async (req, res) => {
+    let config = req.body.loginConfig;
+    let formData = {
+      username: req.body.emailId,
```
Comment (rajeshkumaravel) [inline src/app/routes/keycloakSignInRoutes.js:13]: - What if `req` does not contain `body`
- Exception will be thrown cannot read body of undefined; please add exception or validation for input
- All `req.` reference should be handled for exception

File: src/app/routes/keycloakSignInRoutes.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        logger.info({msg: 'keyclaok Signin route: initialized ' + reqOption.method + reqOption.url})
+        var htmlObject = parser.parseFromString(data, 'text/html');
+        let ele = htmlObject.getElementById("kc-form-login");
+        let loginUrl = ele.getAttribute('action');
+        let loginOption = getRequestOptions(loginUrl, getHeaders(response.headers['set-cookie'].toString(), response.request.href+'?'+response.request.body), formData)
+        request(loginOption, ((e, resp, dataObj) => {
+          if (e) {
+            logger.error({msg: 'keyclaok Signin route : login failed ' + resp.statusCode + dataObj + 'error:'+ JSON.stringify(e)})
+            res.error(e + dataObj);
+          } else if (resp.statusCode == 302 && dataObj == "") {
```
Comment (rajeshkumaravel) [inline src/app/routes/keycloakSignInRoutes.js:36]: - Use `lodash` while referring key from object
- Ex: `_.get(resp, 'statusCode')`

File: src/app/routes/keycloakSignInRoutes.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            let code = resp.headers.location.split('=');
+            let reqTokenOption = getRequestOptions(`${envHelper.PORTAL_AUTH_SERVER_URL}/realms/${envHelper.PORTAL_REALM}/protocol/openid-connect/token`, 
+              getHeaders(), qs.stringify({redirect_uri: redirect_uri, code: code[code.length - 1], grant_type: 'authorization_code',client_id: req.body.client_id}))
+            request(reqTokenOption, ((err, response, data) => {
+              if (!err && response.statusCode == 200) {
+                logger.info({msg: 'keyclaok Signin route : generated access token and refresh token - login sucessful'});
+                res.status(response.statusCode).send(data)
+              } else {
+                logger.error({msg: 'keyclaok Signin route : error while getting access token ' + resp.statusCode + ' error: ' + JSON.parse(data)})
+                res.send(JSON.parse(data));
```
Comment (rajeshkumaravel) [inline src/app/routes/keycloakSignInRoutes.js:47]: - do not use JSON parse without try catch block
- do not log `data`

**Review by rajeshkumaravel — CHANGES_REQUESTED** [review]

File: src/app/routes/keycloakSignInRoutes.js [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+let formData = {
+  username: "",
+  password: ""
+}
+let redirect_uri = "";
+let client_id = "";
+
+module.exports = (app) => {
+  app.post('/keycloak/login', bodyParser.json(), async (req, res) => {
+    if (req && req.body && req.body.loginConfig) {
```
Comment (rajeshkumaravel) [inline src/app/routes/keycloakSignInRoutes.js:19]: `if (req && req.body && req.body.loginConfig)`
can be written as
`if(_.get(req, 'body.loginConfig')`

**Review by rajeshkumaravel — CHANGES_REQUESTED** [review]

File: src/app/routes/keycloakSignInRoutes.js (hunk 6 above)
Comment (vpPavithra) [reply src/app/routes/keycloakSignInRoutes.js:19]: okay

**Review by rajeshkumaravel — APPROVED** [review]

**Review by rajeshkumaravel — APPROVED** [review]

**Review by rajeshkumaravel — APPROVED** [review]

**Review by rajeevsathish — APPROVED** [review]

**Pull request merged**